File: frontend/lib/api/claims.ts
import { api } from "@/lib/api/client";

export type ClaimStatus = "RECEIVED" | "PENDING_REVIEW" | "APPROVED" | "REJECTED";

export type ClaimStatusItem = {
  id: string;
  claim_number: string;
  status: ClaimStatus;
  claimant_name: string;
  billed_amount: number;
  approved_amount: number | null;
  created_at: string;
  updated_at: string;
};

export type ClaimListResponse = {
  total: number;
  items: ClaimStatusItem[];
};

export type SubmitClaimRequest = {
  claim_type: string;
  claimant_id: string;
  claimant_name: string;
  policy_number: string;
  plan_id: string;
  provider_id: string;
  provider_name: string;
  provider_npi?: string;
  facility_name?: string;
  service_date: string;
  icd_codes: string[];
  cpt_codes: string[];
  diagnosis_description: string;
  procedure_description: string;
  billed_amount: number;
  requested_amount: number;
  in_network: boolean;
  prior_auth_number?: string;
  raw_payload: Record<string, unknown>;
};

export type DecisionResponse = {
  status?: "PENDING" | "ERROR";
  message?: string;
  id?: string;
  claim_id?: string;
  verdict?: string;
  confidence?: number;
  approved_amount?: number | null;
  composite_risk_score?: number;
  risk_classification?: string;
  routing_decision?: string;
  fraud_score?: number;
  medical_risk_score?: number;
  policy_risk_score?: number;
  fraud_agent_report?: Record<string, unknown>;
  medical_agent_report?: Record<string, unknown>;
  policy_agent_report?: Record<string, unknown>;
  arbiter_report?: Record<string, unknown>;
  reasoning_tree?: Record<string, unknown>;
  debate_occurred?: boolean;
  debate_transcript?: Record<string, unknown> | null;
  conflict_analysis?: Record<string, unknown> | null;
  human_required?: boolean;
  human_override?: Record<string, unknown> | null;
  denial_reason?: string | null;
  appeals_pathway?: string | null;
  precedent_case_ids?: string[];
  created_at?: string;
  finalized_at?: string | null;
};

export type ClaimWithDecision = {
  claim: ClaimStatusItem;
  decision?: DecisionResponse;
  decisionError?: string;
};

export type DashboardStats = {
  claims: ClaimWithDecision[];
  totalClaims: number;
  approved: number;
  rejected: number;
  pendingReview: number;
  approvalRate: number | null;
  avgRiskScore: number | null;
  totalBilledAmount: number;
  totalApprovedAmount: number;
};

export type AnalyticsData = DashboardStats & {
  claimVolumeByDay: Array<{ date: string; claims: number }>;
  decisionDistribution: Array<{ label: string; value: number; color: string }>;
  riskDistribution: Array<{ label: string; value: number; color: string }>;
  billedVsApproved: Array<{ date: string; billed: number; approved: number }>;
  agentRiskComparison: Array<{ agent: string; risk: number }>;
};

function normalizeVerdict(decision?: DecisionResponse, claim?: ClaimStatusItem): string {
  const raw = decision?.verdict ?? decision?.status ?? claim?.status ?? "PENDING";
  if (raw === "APPROVED") return "APPROVE";
  if (raw === "REJECTED") return "REJECT";
  if (raw === "PENDING_REVIEW") return "REVIEW";
  return raw;
}

function riskScore(decision?: DecisionResponse): number | null {
  const scores = [
    decision?.composite_risk_score,
    decision?.fraud_score,
    decision?.medical_risk_score,
    decision?.policy_risk_score
  ].filter((score): score is number => typeof score === "number");

  if (!scores.length) return null;
  return typeof decision?.composite_risk_score === "number" ? decision.composite_risk_score : Math.max(...scores);
}

function isPendingReview(item: ClaimWithDecision): boolean {
  const verdict = normalizeVerdict(item.decision, item.claim);
  const scores = [item.decision?.fraud_score, item.decision?.medical_risk_score, item.decision?.policy_risk_score].filter(
    (score): score is number => typeof score === "number"
  );

  return (
    verdict === "PENDING" ||
    verdict === "REVIEW" ||
    item.claim.status === "PENDING_REVIEW" ||
    item.decision?.human_required === true ||
    scores.some((score) => score >= 0.55)
  );
}

function dayKey(dateValue?: string): string {
  const date = dateValue ? new Date(dateValue) : new Date();
  if (Number.isNaN(date.getTime())) return "Unknown";
  return date.toLocaleDateString(undefined, { month: "short", day: "numeric" });
}

function average(values: number[]): number | null {
  if (!values.length) return null;
  return values.reduce((sum, value) => sum + value, 0) / values.length;
}

export async function submitClaim(payload: SubmitClaimRequest): Promise<ClaimStatusItem> {
  const { data } = await api.post<ClaimStatusItem>("/claims/submit", payload);
  return data;
}

export async function fetchClaims(status?: string): Promise<ClaimListResponse> {
  return fetchClaimsList(status);
}

export async function fetchClaimsList(status?: string): Promise<ClaimListResponse> {
  const { data } = await api.get<ClaimListResponse>("/claims", {
    params: status ? { status } : undefined
  });
  return data;
}

export async function fetchClaimDetails(claimId: string): Promise<ClaimStatusItem> {
  // Use hardened endpoint (returns 200 w/ PENDING message if not ready)
  const { data } = await api.get<any>(`/claims/${claimId}`);
  if (data?.status === "OK") {
    return {
      id: data.id,
      claim_number: data.claim_number,
      status: data.status_value,
      claimant_name: data.claimant_name,
      billed_amount: data.billed_amount,
      approved_amount: data.approved_amount ?? null,
      created_at: new Date().toISOString(),
      updated_at: new Date().toISOString()
    } as ClaimStatusItem;
  }
  // fallback shape so UI can keep polling
  return {
    id: claimId,
    claim_number: data?.claim_number ?? "",
    status: "PENDING_REVIEW",
    claimant_name: "",
    billed_amount: 0,
    approved_amount: null,
    created_at: new Date().toISOString(),
    updated_at: new Date().toISOString()
  } as ClaimStatusItem;
}

export async function fetchDecision(claimId: string): Promise<DecisionResponse> {
  const { data } = await api.get<DecisionResponse>(`/decisions/${claimId}`);
  return data;
}

export async function fetchClaimsWithDecisions(): Promise<ClaimWithDecision[]> {
  const claimsResponse = await fetchClaims();
  const decisions = await Promise.allSettled(claimsResponse.items.map((claim) => fetchDecision(claim.id)));

  return claimsResponse.items.map((claim, index) => {
    const result = decisions[index];
    if (result.status === "fulfilled") {
      return { claim, decision: result.value };
    }
    return {
      claim,
      decisionError: result.reason instanceof Error ? result.reason.message : "Decision unavailable"
    };
  });
}

export async function fetchDashboardStats(): Promise<DashboardStats> {
  const claims = await fetchClaimsWithDecisions();
  const approved = claims.filter((item) => normalizeVerdict(item.decision, item.claim) === "APPROVE").length;
  const rejected = claims.filter((item) => normalizeVerdict(item.decision, item.claim) === "REJECT").length;
  const pendingReview = claims.filter(isPendingReview).length;
  const riskScores = claims.map((item) => riskScore(item.decision)).filter((score): score is number => score != null);
  const totalBilledAmount = claims.reduce((sum, item) => sum + (item.claim.billed_amount ?? 0), 0);
  const totalApprovedAmount = claims.reduce(
    (sum, item) => sum + (item.decision?.approved_amount ?? item.claim.approved_amount ?? 0),
    0
  );

  return {
    claims,
    totalClaims: claims.length,
    approved,
    rejected,
    pendingReview,
    approvalRate: claims.length ? (approved / claims.length) * 100 : null,
    avgRiskScore: average(riskScores),
    totalBilledAmount,
    totalApprovedAmount
  };
}

export async function fetchAnalytics(): Promise<AnalyticsData> {
  const stats = await fetchDashboardStats();
  const byDay = new Map<string, { date: string; claims: number; billed: number; approved: number }>();

  stats.claims.forEach((item) => {
    const date = dayKey(item.claim.created_at);
    const current = byDay.get(date) ?? { date, claims: 0, billed: 0, approved: 0 };
    current.claims += 1;
    current.billed += item.claim.billed_amount ?? 0;
    current.approved += item.decision?.approved_amount ?? item.claim.approved_amount ?? 0;
    byDay.set(date, current);
  });

  const risks = stats.claims.map((item) => riskScore(item.decision)).filter((score): score is number => score != null);
  const low = risks.filter((score) => score < 0.34).length;
  const medium = risks.filter((score) => score >= 0.34 && score < 0.67).length;
  const high = risks.filter((score) => score >= 0.67).length;

  const fraudScores = stats.claims.map((item) => item.decision?.fraud_score).filter((score): score is number => typeof score === "number");
  const medicalScores = stats.claims
    .map((item) => item.decision?.medical_risk_score)
    .filter((score): score is number => typeof score === "number");
  const policyScores = stats.claims
    .map((item) => item.decision?.policy_risk_score)
    .filter((score): score is number => typeof score === "number");

  return {
    ...stats,
    claimVolumeByDay: Array.from(byDay.values()).map(({ date, claims }) => ({ date, claims })),
    decisionDistribution: [
      { label: "APPROVE", value: stats.approved, color: "#10b981" },
      { label: "PENDING", value: stats.claims.filter((item) => normalizeVerdict(item.decision, item.claim) === "PENDING").length, color: "#a78bfa" },
      { label: "REVIEW", value: stats.pendingReview, color: "#f59e0b" },
      { label: "REJECT", value: stats.rejected, color: "#f43f5e" }
    ],
    riskDistribution: [
      { label: "Low", value: low, color: "#10b981" },
      { label: "Medium", value: medium, color: "#f59e0b" },
      { label: "High", value: high, color: "#ef4444" }
    ],
    billedVsApproved: Array.from(byDay.values()).map(({ date, billed, approved }) => ({ date, billed, approved })),
    agentRiskComparison: [
      { agent: "Fraud", risk: average(fraudScores) ?? 0 },
      { agent: "Medical", risk: average(medicalScores) ?? 0 },
      { agent: "Policy", risk: average(policyScores) ?? 0 }
    ]
  };
}

export const claimDecisionUtils = {
  normalizeVerdict,
  riskScore,
  isPendingReview
};

File: app/api/v1/claims.py
import uuid
from datetime import datetime
from fastapi import APIRouter, Depends, HTTPException, BackgroundTasks
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, func

from app.database import get_db
from app.database import db_available, db_mode
from app.models.claim import Claim, ClaimStatus
from app.schemas.claim import ClaimSubmitRequest, ClaimStatusResponse, ClaimListResponse
from app.utils.helpers import generate_claim_number
from app.utils.audit_logger import emit_event, get_events
from app.schemas.stream import EventTypes
from app.workflow.pipeline import AdjudicationPipeline
from app.models.decision import Decision, HumanReviewTask
from app.in_memory_store import (
    create_claim as create_claim_memory,
    get_claim as get_claim_memory,
    list_claims as list_claims_memory,
    update_claim as update_claim_memory,
    store_decision as store_decision_memory,
)
import structlog

log = structlog.get_logger()
router = APIRouter(prefix="/claims", tags=["Claims"])
_pipeline = AdjudicationPipeline()


async def _run_pipeline(claim_id: str, claim_data: dict) -> None:
    from app.database import AsyncSessionLocal
    log.info("claim_pipeline_started", claim_id=claim_id, db_mode=db_mode())
    try:
        decision_data = await _pipeline.run(claim_data)
        log.info("claim_pipeline_finished", claim_id=claim_id, verdict=decision_data["verdict"])

        if db_available():
            try:
                async with AsyncSessionLocal() as db:
                    # Update claim status
                    claim = await db.get(Claim, claim_id)
                    if claim:
                        if decision_data["verdict"] == "APPROVE":
                            claim.status = ClaimStatus.APPROVED
                            claim.approved_amount = decision_data.get("approved_amount")
                        elif decision_data["verdict"] == "REJECT":
                            claim.status = ClaimStatus.REJECTED
                        else:
                            claim.status = ClaimStatus.PENDING_REVIEW

                    # Persist decision
                    decision = Decision(
                        id=decision_data["id"],
                        claim_id=claim_id,
                        verdict=decision_data["verdict"],
                        confidence=decision_data["confidence"],
                        approved_amount=decision_data.get("approved_amount"),
                        composite_risk_score=decision_data["composite_risk_score"],
                        risk_classification=decision_data["risk_classification"],
                        routing_decision=decision_data["routing_decision"],
                        fraud_score=decision_data["fraud_score"],
                        medical_risk_score=decision_data["medical_risk_score"],
                        policy_risk_score=decision_data["policy_risk_score"],
                        complexity_multiplier=decision_data.get("complexity_multiplier", 1.0),
                        fraud_agent_report=decision_data["fraud_agent_report"],
                        medical_agent_report=decision_data["medical_agent_report"],
                        policy_agent_report=decision_data["policy_agent_report"],
                        arbiter_report=decision_data["arbiter_report"],
                        reasoning_tree=decision_data["reasoning_tree"],
                        debate_occurred=decision_data["debate_occurred"],
                        debate_transcript=decision_data.get("debate_transcript"),
                        conflict_analysis=decision_data.get("conflict_analysis"),
                        human_required=decision_data["human_required"],
                        appeals_pathway=decision_data.get("appeals_pathway"),
                        denial_reason=decision_data.get("denial_reason"),
                        precedent_case_ids=decision_data.get("precedent_case_ids", []),
                        is_simulation=decision_data.get("is_simulation", False),
                        finalized_at=datetime.utcnow() if not decision_data["human_required"] else None,
                    )
                    db.add(decision)

                    # Human review task
                    if decision_data["human_required"]:
                        task = HumanReviewTask(
                            id=str(uuid.uuid4()),
                            claim_id=claim_id,
                            decision_id=decision_data["id"],
                            priority="P1" if decision_data["composite_risk_score"] > 0.85 else "P2",
                            escalation_reason=decision_data.get("denial_reason") or "Risk threshold exceeded",
                            status="PENDING",
                            deadline=datetime.utcnow().replace(hour=0, minute=0, second=0) if False else None,
                        )
                        db.add(task)

                    await db.commit()
                    log.info("decision_persisted", claim_id=claim_id, verdict=decision_data["verdict"], mode="db")
                    return
            except Exception as db_exc:
                log.warning("DB FAILED -> USING MEMORY", claim_id=claim_id, error=str(db_exc))

        next_status = ClaimStatus.PENDING_REVIEW
        approved_amount = None
        if decision_data["verdict"] == "APPROVE":
            next_status = ClaimStatus.APPROVED
            approved_amount = decision_data.get("approved_amount")
        elif decision_data["verdict"] == "REJECT":
            next_status = ClaimStatus.REJECTED

        update_claim_memory(
            claim_id,
            {"status": next_status, "approved_amount": approved_amount},
        )
        store_decision_memory(decision_data)
        log.info("decision_persisted", claim_id=claim_id, verdict=decision_data["verdict"], mode="memory")

    except Exception as e:
        log.error("pipeline_background_error", claim_id=claim_id, error=str(e))
        if db_available():
            async with AsyncSessionLocal() as db:
                claim = await db.get(Claim, claim_id)
                if claim:
                    claim.status = ClaimStatus.PENDING_REVIEW
                    await db.commit()
        else:
            update_claim_memory(claim_id, {"status": ClaimStatus.PENDING_REVIEW})


@router.post("/submit", response_model=ClaimStatusResponse, status_code=202)
async def submit_claim(
    body: ClaimSubmitRequest,
    background_tasks: BackgroundTasks,
    db: AsyncSession = Depends(get_db),
) -> ClaimStatusResponse:
    claim_id = str(uuid.uuid4())
    claim_number = generate_claim_number()
    created_at = datetime.utcnow()
    updated_at = created_at
    try:
        log.info("claim_submit_received", claimant_id=body.claimant_id, provider_id=body.provider_id, db_mode=db_mode())
        claim = Claim(
            id=claim_id,
            claim_number=claim_number,
            status=ClaimStatus.RECEIVED,
            claim_type=body.claim_type,
            claimant_id=body.claimant_id,
            claimant_name=body.claimant_name,
            policy_number=body.policy_number,
            plan_id=body.plan_id,
            provider_id=body.provider_id,
            provider_name=body.provider_name,
            provider_npi=body.provider_npi,
            facility_name=body.facility_name,
            service_date=body.service_date,
            icd_codes=body.icd_codes,
            cpt_codes=body.cpt_codes,
            diagnosis_description=body.diagnosis_description,
            procedure_description=body.procedure_description,
            billed_amount=body.billed_amount,
            requested_amount=body.requested_amount,
            in_network=body.in_network,
            prior_auth_number=body.prior_auth_number,
            raw_payload=body.raw_payload,
        )
        persisted_to_memory = False
        if db_available() and db is not None:
            try:
                db.add(claim)
                await db.commit()
                await db.refresh(claim)
                created_at = claim.created_at
                updated_at = claim.updated_at
                log.info("claim_persisted", claim_id=claim_id, mode="db")
            except Exception as db_exc:
                log.warning("DB FAILED -> USING MEMORY", claim_id=claim_id, error=str(db_exc))
                persisted_to_memory = True
        else:
            persisted_to_memory = True

        if persisted_to_memory:
            create_claim_memory(
                claim_id=claim_id,
                payload={
                    "claim_number": claim_number,
                    "status": ClaimStatus.RECEIVED,
                    "claim_type": body.claim_type,
                    "claimant_id": body.claimant_id,
                    "claimant_name": body.claimant_name,
                    "policy_number": body.policy_number,
                    "plan_id": body.plan_id,
                    "provider_id": body.provider_id,
                    "provider_name": body.provider_name,
                    "provider_npi": body.provider_npi,
                    "facility_name": body.facility_name,
                    "service_date": body.service_date,
                    "icd_codes": body.icd_codes,
                    "cpt_codes": body.cpt_codes,
                    "diagnosis_description": body.diagnosis_description,
                    "procedure_description": body.procedure_description,
                    "billed_amount": body.billed_amount,
                    "requested_amount": body.requested_amount,
                    "in_network": body.in_network,
                    "prior_auth_number": body.prior_auth_number,
                    "raw_payload": body.raw_payload,
                    "approved_amount": None,
                },
            )
            log.info("claim_persisted", claim_id=claim_id, mode="memory")

        emit_event(claim_id, EventTypes.CLAIM_RECEIVED, "ingestion", {"claim_number": claim_number})

        claim_data = {
            "id": claim_id,
            "claim_number": claim_number,
            "claim_type": body.claim_type,
            "claimant_id": body.claimant_id,
            "claimant_name": body.claimant_name,
            "policy_number": body.policy_number,
            "plan_id": body.plan_id,
            "provider_id": body.provider_id,
            "provider_name": body.provider_name,
            "provider_npi": body.provider_npi,
            "facility_name": body.facility_name,
            "service_date": body.service_date.isoformat(),
            "icd_codes": body.icd_codes,
            "cpt_codes": body.cpt_codes,
            "diagnosis_description": body.diagnosis_description,
            "procedure_description": body.procedure_description,
            "billed_amount": body.billed_amount,
            "requested_amount": body.requested_amount,
            "in_network": body.in_network,
            "prior_auth_number": body.prior_auth_number,
            "is_simulation": False,
        }

        background_tasks.add_task(_run_pipeline, claim_id, claim_data)
        log.info("claim_pipeline_enqueued", claim_id=claim_id)
    except Exception as exc:
        log.error("claim_submit_fallback", claim_id=claim_id, error=str(exc))

    return ClaimStatusResponse(
        id=claim_id,
        claim_number=claim_number,
        status=ClaimStatus.RECEIVED,
        claimant_name=body.claimant_name,
        billed_amount=body.billed_amount,
        approved_amount=None,
        created_at=created_at,
        updated_at=updated_at,
    )


@router.get("/{claim_id}/status", response_model=ClaimStatusResponse)
async def get_claim_status(
    claim_id: str,
    db: AsyncSession = Depends(get_db),
) -> ClaimStatusResponse:
    try:
        if db_available() and db is not None:
            claim = await db.get(Claim, claim_id)
            if not claim:
                raise HTTPException(status_code=404, detail="Claim not found")
            return ClaimStatusResponse(
                id=claim.id,
                claim_number=claim.claim_number,
                status=claim.status,
                claimant_name=claim.claimant_name,
                billed_amount=claim.billed_amount,
                approved_amount=claim.approved_amount,
                created_at=claim.created_at,
                updated_at=claim.updated_at,
            )
        claim = get_claim_memory(claim_id)
        if not claim:
            raise HTTPException(status_code=404, detail="Claim not found")
        return ClaimStatusResponse(
            id=claim["id"],
            claim_number=claim["claim_number"],
            status=claim["status"],
            claimant_name=claim["claimant_name"],
            billed_amount=claim["billed_amount"],
            approved_amount=claim.get("approved_amount"),
            created_at=claim["created_at"],
            updated_at=claim["updated_at"],
        )
    except Exception as exc:
        log.warning("claim_status_fallback", claim_id=claim_id, error=str(exc))
        claim = get_claim_memory(claim_id)
        if not claim:
            raise
        return ClaimStatusResponse(
            id=claim["id"],
            claim_number=claim["claim_number"],
            status=claim["status"],
            claimant_name=claim["claimant_name"],
            billed_amount=claim["billed_amount"],
            approved_amount=claim.get("approved_amount"),
            created_at=claim["created_at"],
            updated_at=claim["updated_at"],
        )


@router.get("/{claim_id}", response_model=dict)
async def get_claim(
    claim_id: str,
    db: AsyncSession = Depends(get_db),
) -> dict:
    """
    Hardened claim fetch endpoint for UI.
    Returns 200 with PENDING status when claim not found (prevents UI hard errors).
    """
    try:
        if db_available() and db is not None:
            claim = await db.get(Claim, claim_id)
            if claim:
                return {
                    "status": "OK",
                    "id": claim.id,
                    "claim_number": claim.claim_number,
                    "status_value": str(getattr(claim.status, "value", claim.status)),
                    "claimant_name": claim.claimant_name,
                    "billed_amount": claim.billed_amount,
                    "approved_amount": claim.approved_amount,
                }
        claim = get_claim_memory(claim_id)
        if claim:
            return {
                "status": "OK",
                "id": claim["id"],
                "claim_number": claim["claim_number"],
                "status_value": str(getattr(claim.get("status"), "value", claim.get("status"))),
                "claimant_name": claim["claimant_name"],
                "billed_amount": claim["billed_amount"],
                "approved_amount": claim.get("approved_amount"),
            }
        return {"status": "PENDING", "message": "Decision not ready yet"}
    except Exception as exc:
        log.error("claim_fetch_failed", claim_id=claim_id, error=str(exc))
        return {"status": "PENDING", "message": "Decision not ready yet"}


@router.get("", response_model=ClaimListResponse)
async def list_claims(
    skip: int = 0,
    limit: int = 20,
    status: str | None = None,
    db: AsyncSession = Depends(get_db),
) -> ClaimListResponse:
    if db_available() and db is not None:
        query = select(Claim).offset(skip).limit(limit).order_by(Claim.created_at.desc())
        if status:
            query = query.where(Claim.status == status)

        count_q = select(func.count()).select_from(Claim)
        if status:
            count_q = count_q.where(Claim.status == status)

        result = await db.execute(query)
        count_result = await db.execute(count_q)
        claims = result.scalars().all()
        total = count_result.scalar_one()

        return ClaimListResponse(
            total=total,
            items=[
                ClaimStatusResponse(
                    id=c.id,
                    claim_number=c.claim_number,
                    status=c.status,
                    claimant_name=c.claimant_name,
                    billed_amount=c.billed_amount,
                    approved_amount=c.approved_amount,
                    created_at=c.created_at,
                    updated_at=c.updated_at,
                )
                for c in claims
            ],
        )

    total, claims = list_claims_memory(status=status, skip=skip, limit=limit)
    return ClaimListResponse(
        total=total,
        items=[
            ClaimStatusResponse(
                id=c["id"],
                claim_number=c["claim_number"],
                status=c["status"],
                claimant_name=c["claimant_name"],
                billed_amount=c["billed_amount"],
                approved_amount=c.get("approved_amount"),
                created_at=c["created_at"],
                updated_at=c["updated_at"],
            )
            for c in claims
        ],
    )
